bsp/Src/CAN.c

```diff
@@ -190,156 +191,102 @@ static inline void HAL_CAN_MspF4Init(CAN_HandleTypeDef* hcan){
     PA11 -> CAN1_RX
     PA12 -> CAN1_TX
     */
-    init.Pin = GPIO_PIN_11;
+    init.Pin = GPIO_PIN_11 | GPIO_PIN_12;
     init.Mode = GPIO_MODE_AF_PP;
     init.Pull = GPIO_PULLUP;
     init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
     init.Alternate = GPIO_AF9_CAN1;
     HAL_GPIO_Init(GPIOA, &init);
-
-    init.Pin = GPIO_PIN_12;
-    init.Mode = GPIO_MODE_AF_PP;
-    init.Pull = GPIO_NOPULL;
-    init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
-    init.Alternate = GPIO_AF9_CAN1;
-    HAL_GPIO_Init(GPIOA, &init);
+    return;
   }
+  #endif
 
   // CAN2
   #ifdef CAN2
-  else if (hcan->Instance == CAN2) {
+  if (hcan->Instance == CAN2) {
     // enable clocks
     __HAL_RCC_GPIOB_CLK_ENABLE();
 
     /* enable gpio
     PB12 -> CAN2_RX
     PB13 -> CAN2_TX
     */
-    init.Pin = GPIO_PIN_12;
+    init.Pin = GPIO_PIN_12 | GPIO_PIN_13;
     init.Mode = GPIO_MODE_AF_PP;
     init.Pull = GPIO_PULLUP;
     init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
     init.Alternate = GPIO_AF9_CAN2;
     HAL_GPIO_Init(GPIOB, &init);
-
-    init.Pin = GPIO_PIN_13;
-    init.Mode = GPIO_MODE_AF_PP;
-    init.Pull = GPIO_NOPULL;
-    init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
-    init.Alternate = GPIO_AF9_CAN2;
-    HAL_GPIO_Init(GPIOB, &init);
+    return;
   }
   #endif /* CAN2 */
-
+  
   // CAN3
   #ifdef CAN3
-  else if (hcan->Instance == CAN3) {
+  if (hcan->Instance == CAN3) {
     // enable clocks
     __HAL_RCC_GPIOA_CLK_ENABLE();
 
     /* enable gpio
     PA8  -> CAN3_RX
     PA15 -> CAN3_TX
     */
-    init.Pin = GPIO_PIN_8;
+    init.Pin = GPIO_PIN_8 | GPIO_PIN_15;
     init.Mode = GPIO_MODE_AF_PP;
     init.Pull = GPIO_PULLUP;
     init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
     init.Alternate = GPIO_AF11_CAN3;
     HAL_GPIO_Init(GPIOA, &init);
-
-    init.Pin = GPIO_PIN_15;
-    init.Mode = GPIO_MODE_AF_PP;
-    init.Pull = GPIO_NOPULL;
-    init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
-    init.Alternate = GPIO_AF11_CAN3;
-    HAL_GPIO_Init(GPIOA, &init);
+    return;
   }
   #endif /* CAN3 */
 }
 
+// CAN GPIO deinit, can be overridden by user
+__weak void HAL_CAN_MspGPIODeInit(CAN_HandleTypeDef* hcan){
+  return;
+}
```

Comment:
does this need to have stuff in the function def to deinit the stuff for the default behavior? in other words, if the user doesn't define MspGPIOInit, it will be the default behavior for GPIOInit, so then the GPIODeInit should deinit that in default behavior as well?